File: lib/factorix/ser_des/deserializer.rb
# frozen_string_literal: true

module Factorix
  module SerDes
    # Deserialize data from binary format
    #
    # This class provides methods to deserialize various data types from Factorio's
    # binary file format, following the specifications documented in the Factorio wiki.
    class Deserializer
      # @!parse
      #   # @return [Dry::Logger::Dispatcher]
      #   attr_reader :logger
      include Import[:logger]

      # Create a new Deserializer instance
      #
      # @param stream [IO] An IO-like object that responds to #read
      # @param logger [Dry::Logger::Dispatcher] optional logger
      # @raise [ArgumentError] If the stream doesn't respond to #read
      def initialize(stream, logger: nil)
        super(logger:)
        raise ArgumentError, "can't read from the given argument" unless stream.respond_to?(:read)

        @stream = stream
        logger.debug "Initializing Deserializer"
      end

      # Read raw bytes from the stream
      #
      # @param length [Integer] Number of bytes to read
      # @raise [Factorix::InvalidLengthError] If length is nil or negative
      # @raise [EOFError] If end of file is reached before reading length bytes
      # @return [String] Binary data read
      def read_bytes(length)
        raise InvalidLengthError, "nil length" if length.nil?
        raise InvalidLengthError, "negative length #{length}" if length.negative?
        return +"" if length.zero?

        bytes = @stream.read(length)
        if bytes.nil? || bytes.size < length
          logger.debug("Unexpected EOF", requested_bytes: length)
          raise EOFError
        end

        bytes
      end

      # Read an unsigned 8-bit integer
      #
      # @return [Integer] 8-bit unsigned integer
      def read_u8 = read_bytes(1).unpack1("C")

      # Read an unsigned 16-bit integer
      #
      # @return [Integer] 16-bit unsigned integer
      def read_u16 = read_bytes(2).unpack1("v")

      # Read an unsigned 32-bit integer
      #
      # @return [Integer] 32-bit unsigned integer
      def read_u32 = read_bytes(4).unpack1("V")

      # Read a space-optimized 16-bit unsigned integer
      #
      # @see https://wiki.factorio.com/Data_types#Space_Optimized
      # @return [Integer] 16-bit unsigned integer
      def read_optim_u16
        byte = read_u8
        byte == 0xFF ? read_u16 : byte
      end

      # Read a space-optimized 32-bit unsigned integer
      #
      # @see https://wiki.factorio.com/Data_types#Space_Optimized
      # @return [Integer] 32-bit unsigned integer
      def read_optim_u32
        byte = read_u8
        byte == 0xFF ? read_u32 : byte
      end

      # Read a tuple of 16-bit unsigned integers
      #
      # @param length [Integer] Number of integers to read
      # @return [Array<Integer>] Array of 16-bit unsigned integers
      def read_u16_tuple(length) = Array.new(length) { read_u16 }

      # Read a boolean value
      #
      # @return [Boolean] Boolean value
      def read_bool = read_u8 != 0

      # Read a string
      #
      # @return [String] String read
      def read_str
        length = read_optim_u32
        read_bytes(length).force_encoding(Encoding::UTF_8)
      end

      # Read a string property
      #
      # @see https://wiki.factorio.com/Property_tree#String
      # @return [String] String property
      def read_str_property = read_bool ? "" : read_str

      # Read a double-precision floating point number
      #
      # @see https://wiki.factorio.com/Property_tree#Number
      # @return [Float] Double-precision floating point number
      def read_double = read_bytes(8).unpack1("d")

      # Read a GameVersion object
      #
      # @return [GameVersion] GameVersion object
      def read_game_version = Types::GameVersion.from_numbers(read_u16, read_u16, read_u16, read_u16)

      # Read a MODVersion object
      #
      # @return [MODVersion] MODVersion object
      def read_mod_version = Types::MODVersion.from_numbers(read_optim_u16, read_optim_u16, read_optim_u16)

      # Read a signed long integer (8 bytes)
      #
      # @return [Integer] Signed long integer
      def read_long = read_bytes(8).unpack1("q<")

      # Read an unsigned long integer (8 bytes)
      #
      # @return [Integer] Unsigned long integer
      def read_unsigned_long = read_bytes(8).unpack1("Q<")

      # Read a dictionary
      #
      # @see https://wiki.factorio.com/Property_tree#Dictionary
      # @return [Hash] Dictionary of key-value pairs
      def read_dictionary
        length = read_u32
        logger.debug("Reading dictionary", length:)
        length.times.each_with_object({}) do |_i, dict|
          key = read_str_property
          dict[key] = read_property_tree
        end
      end

      # Read a list
      # This type is identical to dictionary
      #
      # @see https://wiki.factorio.com/Property_tree#List
      alias read_list read_dictionary

      # Read a property tree
      #
      # @raise [Factorix::UnknownPropertyType] If the property type is not supported
      # @return [Object] Object read
      def read_property_tree
        type = read_u8
        _any_type_flag = read_bool
        logger.debug("Reading property tree", type:)

        case type
        when 0
          # Handle type 0 - None (null value)
          #
          # @see https://wiki.factorio.com/Property_tree
          nil
        when 1
          read_bool
        when 2
          read_double
        when 3
          read_str_property
        when 4
          read_list
        when 5
          read_dictionary
        when 6
          # Handle type 6 - Signed integer
          #
          # @see https://wiki.factorio.com/Property_tree
          SignedInteger.new(read_long)
        when 7
          # Handle type 7 - Unsigned integer
          #
          # @see https://wiki.factorio.com/Property_tree
          UnsignedInteger.new(read_unsigned_long)
        else
          logger.debug("Unknown property type", type:)
          raise UnknownPropertyType, "Unknown property type: #{type}"
        end
      end

      # Check if the stream is at EOF
      #
      # @return [Boolean] True if at end of file, false otherwise
      def eof? = @stream.eof?
    end
  end
end

File: internal/serdes/deserializer.go
package serdes

import (
	"encoding/binary"
	"fmt"
	"io"
	"math"

	"github.com/sakuro/factorix/internal/mod"
)

// Deserializer reads Factorio's binary format from a stream.
type Deserializer struct {
	r io.Reader
}

// NewDeserializer wraps r for reading.
func NewDeserializer(r io.Reader) *Deserializer {
	return &Deserializer{r: r}
}

// ReadBytes reads exactly n bytes, failing with io.ErrUnexpectedEOF (or
// io.EOF when nothing was read) on a short stream.
func (d *Deserializer) ReadBytes(n int) ([]byte, error) {
	if n < 0 {
		return nil, fmt.Errorf("%w: %d", ErrInvalidLength, n)
	}
	buf := make([]byte, n)
	if _, err := io.ReadFull(d.r, buf); err != nil {
		return nil, err
	}
	return buf, nil
}

// ReadU8 reads an unsigned 8-bit integer.
func (d *Deserializer) ReadU8() (uint8, error) {
	b, err := d.ReadBytes(1)
	if err != nil {
		return 0, err
	}
	return b[0], nil
}

// ReadU16 reads a little-endian unsigned 16-bit integer.
func (d *Deserializer) ReadU16() (uint16, error) {
	b, err := d.ReadBytes(2)
	if err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint16(b), nil
}

// ReadU32 reads a little-endian unsigned 32-bit integer.
func (d *Deserializer) ReadU32() (uint32, error) {
	b, err := d.ReadBytes(4)
	if err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint32(b), nil
}

// ReadOptimU16 reads a space-optimized 16-bit unsigned integer.
//
// See https://wiki.factorio.com/Data_types#Space_Optimized
func (d *Deserializer) ReadOptimU16() (uint16, error) {
	b, err := d.ReadU8()
	if err != nil {
		return 0, err
	}
	if b == 0xFF {
		return d.ReadU16()
	}
	return uint16(b), nil
}

// ReadOptimU32 reads a space-optimized 32-bit unsigned integer.
//
// See https://wiki.factorio.com/Data_types#Space_Optimized
func (d *Deserializer) ReadOptimU32() (uint32, error) {
	b, err := d.ReadU8()
	if err != nil {
		return 0, err
	}
	if b == 0xFF {
		return d.ReadU32()
	}
	return uint32(b), nil
}

// ReadBool reads a boolean value.
func (d *Deserializer) ReadBool() (bool, error) {
	b, err := d.ReadU8()
	if err != nil {
		return false, err
	}
	return b != 0, nil
}

// ReadStr reads a length-prefixed string.
func (d *Deserializer) ReadStr() (string, error) {
	length, err := d.ReadOptimU32()
	if err != nil {
		return "", err
	}
	b, err := d.ReadBytes(int(length))
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// ReadStrProperty reads a string preceded by an "is empty" flag.
//
// See https://wiki.factorio.com/Property_tree#String
func (d *Deserializer) ReadStrProperty() (string, error) {
	empty, err := d.ReadBool()
	if err != nil {
		return "", err
	}
	if empty {
		return "", nil
	}
	return d.ReadStr()
}

// ReadDouble reads a little-endian double-precision floating point number.
func (d *Deserializer) ReadDouble() (float64, error) {
	b, err := d.ReadBytes(8)
	if err != nil {
		return 0, err
	}
	return math.Float64frombits(binary.LittleEndian.Uint64(b)), nil
}

// ReadLong reads a little-endian signed 64-bit integer.
func (d *Deserializer) ReadLong() (int64, error) {
	v, err := d.ReadUnsignedLong()
	return int64(v), err
}

// ReadUnsignedLong reads a little-endian unsigned 64-bit integer.
func (d *Deserializer) ReadUnsignedLong() (uint64, error) {
	b, err := d.ReadBytes(8)
	if err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint64(b), nil
}

// ReadGameVersion reads a GameVersion (4 x u16).
func (d *Deserializer) ReadGameVersion() (mod.GameVersion, error) {
	var parts [4]uint16
	for i := range parts {
		v, err := d.ReadU16()
		if err != nil {
			return mod.GameVersion{}, err
		}
		parts[i] = v
	}
	return mod.GameVersion{Major: parts[0], Minor: parts[1], Patch: parts[2], Build: parts[3]}, nil
}

// ReadMODVersion reads a MODVersion (3 x optim u16).
func (d *Deserializer) ReadMODVersion() (mod.MODVersion, error) {
	var parts [3]uint16
	for i := range parts {
		v, err := d.ReadOptimU16()
		if err != nil {
			return mod.MODVersion{}, err
		}
		parts[i] = v
	}
	return mod.NewMODVersion(parts[0], parts[1], parts[2])
}

// ReadList reads a property tree list. The binary structure is identical to
// a dictionary; the per-item keys are empty and discarded.
//
// See https://wiki.factorio.com/Property_tree#List
func (d *Deserializer) ReadList() ([]PropertyTree, error) {
	count, err := d.ReadU32()
	if err != nil {
		return nil, err
	}
	var items []PropertyTree
	for range count {
		if _, err := d.ReadStrProperty(); err != nil {
			return nil, err
		}
		item, err := d.ReadPropertyTree()
		if err != nil {
			return nil, err
		}
		items = append(items, item)
	}
	return items, nil
}

// ReadDictionary reads a property tree dictionary, preserving entry order.
//
// See https://wiki.factorio.com/Property_tree#Dictionary
func (d *Deserializer) ReadDictionary() ([]DictEntry, error) {
	count, err := d.ReadU32()
	if err != nil {
		return nil, err
	}
	var entries []DictEntry
	for range count {
		key, err := d.ReadStrProperty()
		if err != nil {
			return nil, err
		}
		value, err := d.ReadPropertyTree()
		if err != nil {
			return nil, err
		}
		entries = append(entries, DictEntry{Key: key, Value: value})
	}
	return entries, nil
}

// ReadPropertyTree reads one property tree element.
func (d *Deserializer) ReadPropertyTree() (PropertyTree, error) {
	typ, err := d.ReadU8()
	if err != nil {
		return PropertyTree{}, err
	}
	// The "any-type" flag is unused by Factorio; read and discard.
	if _, err := d.ReadBool(); err != nil {
		return PropertyTree{}, err
	}

	switch Kind(typ) {
	case KindNone:
		return None(), nil
	case KindBool:
		v, err := d.ReadBool()
		if err != nil {
			return PropertyTree{}, err
		}
		return Bool(v), nil
	case KindNumber:
		v, err := d.ReadDouble()
		if err != nil {
			return PropertyTree{}, err
		}
		return Number(v), nil
	case KindString:
		v, err := d.ReadStrProperty()
		if err != nil {
			return PropertyTree{}, err
		}
		return String(v), nil
	case KindList:
		items, err := d.ReadList()
		if err != nil {
			return PropertyTree{}, err
		}
		return List(items...), nil
	case KindDict:
		entries, err := d.ReadDictionary()
		if err != nil {
			return PropertyTree{}, err
		}
		return Dict(entries...), nil
	case KindSignedInt:
		v, err := d.ReadLong()
		if err != nil {
			return PropertyTree{}, err
		}
		return SignedInt(v), nil
	case KindUnsignedInt:
		v, err := d.ReadUnsignedLong()
		if err != nil {
			return PropertyTree{}, err
		}
		return UnsignedInt(v), nil
	default:
		return PropertyTree{}, &UnknownPropertyTypeError{Type: typ}
	}
}
